src/Common/SIEVECachePolicy.h

```diff
@@ -0,0 +1,212 @@
+#pragma once
+
+#include <Common/ICachePolicy.h>
+
+#include <list>
+#include <unordered_map>
+
+namespace DB
+{
+/// Cache policy SIEVE evicts entries which are not used for a long time. Also see cache policy SLRU for reference.
+/// WeightFunction is a functor that takes Mapped as a parameter and returns "weight" (approximate size) of that value.
+/// Cache starts to evict entries when their total weight exceeds max_size_in_bytes.
+/// Value weight should not change after insertion.
+/// To work with the thread-safe implementation of this class use a class "CacheBase" with first parameter "SIEVE"
+/// and next parameters in the same order as in the constructor of the current class.
+template <typename Key, typename Mapped, typename HashFunction = std::hash<Key>, typename WeightFunction = EqualWeightFunction<Mapped>>
+class SIEVECachePolicy : public ICachePolicy<Key, Mapped, HashFunction, WeightFunction>
```

Comment:
Please add unit tests which checks correctness of the algorithm in `src/Common/tests/` similar to https://github.com/ClickHouse/ClickHouse/blob/master/src/Common/tests/gtest_slru_cache.cpp, (to build run `ninja unit_tests_dbms`, then you can run it: `build_dir/src/unit_tests_dbms --gtest_filter='*test_name*'`).